server.go

```diff
@@ -1103,20 +1103,38 @@ func chainUnaryServerInterceptors(s *Server) {
 }
 
 func chainUnaryInterceptors(interceptors []UnaryServerInterceptor) UnaryServerInterceptor {
+	// the struct ensures the variables are allocated together, rather than separately, since we
+	// know they should be garbage collected together. This, together with the sync.Pool, saves
+	// 2 allocation and decreases time/call by about 25% on the microbenchmark.
+	type state struct {
+		i            int
+		next         UnaryHandler
+		info         *UnaryServerInfo
+		handler      UnaryHandler
+		interceptors []UnaryServerInterceptor
```

Comment:
Why does `state` need this?  Can't we access `interceptors` directly instead?

I also don't understand the inclusion of `info` and `handler` here.  How does this help?

Reply:
If they are not included, as they were before, the lambda captures them by allocating space for them on the heap.
By including them, the space for these "captures" is provided by the struct itself, that we reuse via Pool, thereby avoiding the corresponding allocations of these lambda captures. This can be trivially shown by removing those fields and noticing that benchmem reports additional allocated space.